src/System.Private.CoreLib/shared/System/Globalization/TextInfo.cs

```diff
@@ -373,49 +383,142 @@ private unsafe string ChangeCase(string source, bool toUpper)
             return result;
         }
 
-        internal unsafe void ChangeCase(ReadOnlySpan<char> source, Span<char> destination, bool toUpper)
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal void ChangeCaseToLower(ReadOnlySpan<char> source, Span<char> destination)
         {
-            Debug.Assert(!_invariantMode);
             Debug.Assert(destination.Length >= source.Length);
+            ChangeCaseCommon<ToLowerConversion>(ref MemoryMarshal.GetReference(source), ref MemoryMarshal.GetReference(destination), source.Length);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal void ChangeCaseToUpper(ReadOnlySpan<char> source, Span<char> destination)
+        {
+            Debug.Assert(destination.Length >= source.Length);
+            ChangeCaseCommon<ToUpperConversion>(ref MemoryMarshal.GetReference(source), ref MemoryMarshal.GetReference(destination), source.Length);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void ChangeCaseCommon<TConversion>(ReadOnlySpan<char> source, Span<char> destination) where TConversion : struct
+        {
+            Debug.Assert(destination.Length >= source.Length);
+            ChangeCaseCommon<TConversion>(ref MemoryMarshal.GetReference(source), ref MemoryMarshal.GetReference(destination), source.Length);
+        }
+
+        private void ChangeCaseCommon<TConversion>(ref char source, ref char destination, int charCount) where TConversion : struct
+        {
+            Debug.Assert(typeof(TConversion) == typeof(ToUpperConversion) || typeof(TConversion) == typeof(ToLowerConversion));
+            bool toUpper = typeof(TConversion) == typeof(ToUpperConversion); // JIT will treat this as a constant in release builds
 
-            if (source.IsEmpty)
+            Debug.Assert(!_invariantMode);
+            Debug.Assert(charCount >= 0);
+
+            if (charCount == 0)
             {
-                return;
+                goto Return;
             }
 
-            fixed (char* pSource = &MemoryMarshal.GetReference(source))
-            fixed (char* pResult = &MemoryMarshal.GetReference(destination))
```

Comment:
Is the minor saving from avoiding the pinning really worth it here? The byref-arithmetic is very hard to read.

Reply:
It doesn't affect benchmark perf significantly. This is mainly to make the GC happier by keeping fewer objects pinned just in case a GC occurs.

Is the concern the syntax? We could add an internal method similar to the following, which would make the call sites look much nicer, while minimizing the number of pinned objects.

```cs
// new internal API on Unsafe class
static internal ref T ReadUnaligned<T>(ref T @base, nuint index, nuint displacementBytes) { /* ... */ }

// old call site
// mov tmp, dword ptr [source + 2 * currIdx + 4]
tempValue = Unsafe.ReadUnaligned<uint>(ref Unsafe.As<char, byte>(ref Unsafe.AddByteOffset(ref Unsafe.Add(ref source, (nint)currIdx), 4)));

// new call site
// mov tmp, dword ptr [source + 2 * currIdx + 4]
tempValue = Unsafe.ReadUnaligned<char, uint>(ref source, (nint)currIdx, 4);
```

Alternatively, if we're ok with the GC impact of pinning objects more than necessary, we can just go back to raw pointers.

Reply:
> This is mainly to make the GC happier by keeping fewer objects pinned just in case a GC occurs.

Short-term pinning is not really a problem for the GC. We have it everywhere - the GC is tuned to deal with it well.

Reply:
As expected, the benchmark doesn't change when we go back to using pointers.

I spoke briefly with Maoni and asked for guidance on when it would make sense to use pointers vs. when it would make sense to avoid pinning. She gave a good rule of thumb: try to avoid pinning if you're going to be running your logic for an extended period of time, and try to avoid pinning if you're working with potentially large data sets.

Since this method is quick and since we expect input strings to be small, I agree with your assessment that avoiding pinning doesn't really buy us much. Will revert.